L1Trigger/L1THGCal/src/backend_emulator/HGCalHistoClusteringImpl_SA.cc

```diff
@@ -0,0 +1,807 @@
+  vector<unsigned int> clock( config_.cColumns(), 0 );
+  for ( const auto& protoCluster : protoClusters ) {
+    protoClustersPerColumn.at( protoCluster->index() ).push_back( protoCluster );
+  }
+
+  map<unsigned int, HGCalClusterSAPtr> sums;
+
+  for ( const auto& flag : readoutFlags ) {
+    auto accumulator = make_shared<HGCalCluster>( 0,
+                                                  0,
+                                                  true, true
+                                                );
+    flag->setClock( flag->clock() + 23 ); // Magic numbers
+
+    for ( const auto& protoCluster : protoClustersPerColumn.at( flag->index() ) ) {
+      if ( protoCluster->clock() <= clock.at( flag->index() ) ) continue;
+      if ( protoCluster->clock() > flag->clock() ) continue;
+      *accumulator += *protoCluster;
+    }
+
+    clock.at( flag->index() ) = flag->clock();
+    accumulator->setClock( flag->clock() );
+    accumulator->setIndex( flag->index() );
+    accumulator->setDataValid( true );
+    clusterAccumulation.push_back( accumulator );
+
+    if ( sums.find( flag->clock() ) == sums.end() ) {
+      auto sum = make_shared<HGCalCluster>( flag->clock() + 7, // Magic numbers
+                                            0,
+                                            true, true
+                                          );
+      sums[flag->clock()] = sum;
+    }
+
+    *(sums.at( flag->clock() )) += *accumulator;
+  }
+
+  for (const auto& sum: sums) {
+    clusterSums.push_back( sum.second );
+  }
+
+  // std::cout << "Output from ClusterSum" << std::endl;
+  // unsigned int nTCs = 0;
+  // for ( const auto& c : clusterAccumulation ) {
+  //   nTCs += c->n_tc();
+  // }
+  // std::cout << nTCs << std::endl;
+  // nTCs = 0;
+  //unsigned int ID = 0;
+  // Print the info as the Python emu format
+  //std::cout << "ID,event_ID,cluster_ID,N_TC,E,E_EM,E_EM_core,E_H_early,W,N_TC_W,W2,Wz,Weta,Wphi,Wroz,Wz2,Weta2,Wphi2,Wroz2,LayerBits,Sat_TC,ShapeQ,SortKey" << std::endl;
+  //for ( const auto& c : clusterSums ) {
+     //std::cout << ID << ",0," << c->index() << "," << c->n_tc() << "," << c->e() << "," << c->e_em() << "," << c->e_em_core() << "," << c->e_h_early() << "," << c->w() << "," << c->n_tc_w() << "," << c->w2() << "," << c->wz() << "," << c->weta() << "," << c->wphi() << "," << c->wroz() << "," << c->wz2() << "," << c->weta2() << "," << c->wphi2() << "," << c->wroz2() << "," << c->layerbits() << "," << c->sat_tc() << "," << c->shapeq() << ",0" << std::endl;
+     //ID++;
+     //nTCs += c->n_tc();
+  //}
+  // std::cout << nTCs << std::endl;
+
+
+}
+
+std::pair< unsigned int, unsigned int > HGCalHistoClusteringImplSA::sigma_Energy(unsigned int N_TC_W, unsigned long int Sum_W2, unsigned int Sum_W) const {
+  unsigned long int N = N_TC_W*Sum_W2 - pow(Sum_W,2);
+  unsigned long int D = pow(N_TC_W,2);
+  double intpart;
+  double frac =  modf(sqrt(N/D),&intpart)*pow(2,1);
```

Comment:
@gfedi After rebasing my changes, the code crashes here.  We need to protect against cases where N_TC_W = 0, and in other places e.g. Sum_W = 0 in `mean_coorindate` etc.

The current clustering emulation can output clusters with 0 weight/constituents, which may change in the future, but I guess we should be careful here in any case.  Perhaps we avoid calculating properties for any cluster with 0 weight/consitutents?

Reply:
Addressed in https://github.com/PFCal-dev/cmssw/pull/499/commits/8a48a5bb8251b01b8d24a962964a376c64e4a6c5